README.md
# tilelive-blend

Will it blend?

## Usage

```javascript
{
  protocol: "blend:",
  query: {
    layers: [
      {
        source: "tilejson+http://tile.stamen.com/toner-background/index.json",
        opacity: 0.5,
        filters: "color-to-alpha(#008800)",
        offset: [5, -5]
      },
      {
        source: "tilejson+http://tile.stamen.com/toner-lines/index.json",
        filters: "invert agg-stack-blur(1,1)"
      },
      {
        source: "tilejson+http://tile.stamen.com/toner-labels/index.json",
        "comp-op": "over"
      }
    ],
    format: "png32:z=1"
  }
}
```

If you need to blend over a solid color, use
[`tilelive-solid`](https://github.com/mojodna/tilelive-solid).


CHANGELOG.md
# Changes

## v0.5.1 - 6/14/17

* Fix dependency `require`s

## v0.5.0 - 6/13/17

* Update dependencies

## v0.4.6 - 7/10/16

* Include the correct number of params when calling back

## v0.4.5 - 7/10/16

* Call the correct callback when a tile is out of range

## v0.4.4 - 7/10/16

* Actually, those should have been errors

## v0.4.3 - 7/10/16

* Include the correct number of params when calling back

## v0.4.2 - 7/10/16

* Handle layer URIs passed as objects

## v0.4.1 - 7/10/16

* Default `query` property of `uri`

## v0.4.0 - 7/10/16

* Allow `info` to be provided in `uri.query`
* Remove `layers` and `format` from exposed `info`

## v0.3.1 - 7/7/16

* Invalidate tiles that experienced errors, not missing tiles

## v0.3.0 - 7/6/16

* Calculate an appropriate `Cache-Control` header

## v0.2.3 - 7/6/16

* Default `scale` to `1`

## v0.2.2 - 7/6/16

* Calculate tile size according to scale (if not provided)

## v0.2.1 - 7/5/16

* Emit proper headers according to output image type

## v0.2.0 - 2/22/16

* Update dependencies
* Recast configuration with stacks
* Only request tiles expected to exist
* Log errors raised when loading sources
* Convert `mapnik` dependency to a peer dependency

## v0.1.0 - 7/11/14

* Initial version


index.js
"use strict";

var url = require("url"),
    util = require("util");

var async = require("async"),
    mapnik = require("mapnik"),
    mercator = new (require("@mapbox/sphericalmercator"))(),
    tiletype = require("@mapbox/tiletype");

var OPS = require("./comp-ops.json");

/**
 * Find the shortest provided max-age Cache-Control header in a list of
 * responses.
 */
var findShortestMaxAge = function(headers) {
  // determine the shortest max-age for an upstream tile and use that
  var maxAges = headers.map(function(h) {
    return (h["Cache-Control"] || h["cache-control"] || "").split(",").map(function(c) {
      return c.trim();
    }).filter(function(c) {
      return c.match(/^max-age=/);
    }).map(function(c) {
      return +c.split("=")[1];
    })[0];
  }).filter(function(age) {
    return age != null;
  });

  return Math.min.apply(null, maxAges);
};

module.exports = function(tilelive, options) {
  var Blend = function(uri, callback) {
    if (typeof(uri) === "string") {
      uri = url.parse(uri, true);
    }

    this.layers = uri.query.layers || [];
    this.format = uri.query.format || "png32";
    this.scale = +uri.query.scale || 1;
    this.tileSize = (uri.query.tileSize | 0) || Math.floor(256 * this.scale);

    this.info = uri.query.info || {};

    return setImmediate(callback, null, this);
  };

  // TODO allow custom headers (User-Agent, X-Forwarded-For) to be passed
  // through
  Blend.prototype.getTile = function(z, x, y, callback) {
    var layers = this.layers,
        format = this.format,
        scale = this.scale,
        tileSize = this.tileSize;

    return async.map(this.layers, function(layer, done) {
      return async.waterfall([
        function(cb) {
          if (typeof(layer.source) === "string") {
            layer = url.parse(layer.source, true);
          } else {
            layer = layer.source;
          }

          layer.query = layer.query || {};

          layer.query.scale = layer.query.scale || scale;
          layer.query.tileSize = layer.query.tileSize || tileSize;

          return tilelive.load(layer, cb);
        },
        function(source, cb) {
          return async.waterfall([
            function(_cb) {
              return source.getInfo(_cb);
            },
            function(info, _cb) {
              if (z < Math.max(0, info.minzoom | 0) || z > (info.maxzoom || Infinity)) {
                return _cb(null, null, {});
              }

              var xyz = mercator.xyz(info.bounds || [-180, -85.0511, 180, 85.0511], z);

              if (x < xyz.minX ||
                  x > xyz.maxX ||
                  y < xyz.minY ||
                  y > xyz.maxY) {
                return _cb(null, null, {});
              }

              return source.getTile(z, x, y, function(err, buffer, headers) {
                headers = headers || {};

                if (err && !err.message.match(/(Tile|Grid) does not exist/)) {
                  console.warn(err.stack);

                  // mark this tile as needing to be invalidated immediately
                  headers["cache-control"] = "public, max-age=0";
                }

                // always claim success; it'll be treated as a transparent tile
                // if it failed
                return _cb(null, buffer, headers);
              });
            }
          ], cb);
        },
        function(buffer, headers, cb) {
          if (buffer) {
            return mapnik.Image.fromBytes(buffer, function(err, data) {
              return cb(err, data, headers);
            });
          }

          // create an empty image if no data was provided
          return cb(null, new mapnik.Image(tileSize, tileSize), headers);
        },
        function(im, headers, cb) {
          return im.premultiply(function(err) {
            return cb(err, [im, headers]);
          });
        }
      ], done);
    }, function(err, data) {
      if (err) {
        console.warn(err);
        return callback(err);
      }

      var images = data.map(function(x) {
          return x[0];
        }),
        headers = data.map(function(x) {
          return x[1];
        }),
        maxAge = findShortestMaxAge(headers);

      var idx = 0;

      return async.waterfall([
        function(done) {
          return async.reduce(images, new mapnik.Image(tileSize, tileSize, {
            premultiplied: true
          }), function(im1, im2, cb) {
            return im1.composite(im2, {
              comp_op: mapnik.compositeOp[OPS[(layers[idx] || {})["comp-op"]] || "src_over"],
              opacity: parseFloat(layers[idx].opacity || 1),
              image_filters: layers[idx].filters || "",
              dx: Number((layers[idx].offset || [])[0] || 0),
              dy: Number((layers[idx++].offset || [])[1] || 0)
            }, cb);
          }, done);
        },
        function(im, cb) {
          return im.demultiply(cb);
        },
        function(im, cb) {
          return im.encode(format, cb);
        }
      ], function(err, buffer) {
        if (err) {
          return callback(err);
        }

        var headers = tiletype.headers(buffer);

        if (maxAge != null && maxAge < Infinity) {
          headers["cache-control"] = util.format("public, max-age=%d", maxAge);
        }

        return callback(null, buffer, headers);
      });
    });
  };

  Blend.prototype.getInfo = function(callback) {
    return setImmediate(callback, null, this.info);
  };

  Blend.prototype.close = function(callback) {
    return callback && setImmediate(callback);
  };

  Blend.registerProtocols = function(tilelive) {
    tilelive.protocols["blend:"] = Blend;
  };

  Blend.registerProtocols(tilelive);

  return Blend;
};


example.js
"use strict";

var fs = require("fs");

var tilelive = require("@mapbox/tilelive");

require("@mapbox/tilejson").registerProtocols(tilelive);
require("./")(tilelive);

tilelive.load({
  protocol: "blend:",
  query: {
    layers: [
      {
        source: "tilejson+http://staging.tile.stamen.com/toner-background/index.json",
        opacity: 0.5,
        filters: "color-to-alpha(#008800)",
        offset: [5, -5]
      },
      {
        source: "tilejson+http://staging.tile.stamen.com/toner-lines/index.json",
        filters: "invert agg-stack-blur(1,1)" 
      },
      {
        source: "tilejson+http://staging.tile.stamen.com/toner-labels/index.json",
        "comp-op": "over"
      }
    ],
    format: "png32:z=1"
  }
}, function(err, source) {
  if (err) {
    throw err;
  }

  return source.getTile(12, 656, 1430, function(err, data) {
    if (err) {
      throw err;
    }

    return fs.writeFileSync("./out.png", data);
  });
});
